clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -635,6 +639,9 @@ public void handle(SyncGroupResponse syncResponse,
                 } else if (error == Errors.REBALANCE_IN_PROGRESS) {
                     log.debug("SyncGroup failed because the group began another rebalance");
                     future.raise(error);
+                } else if (error == Errors.FENCED_INSTANCE_ID) {
```

Comment:
Similar to commit, if this error is thrown, then in `AbstractCoordinator#initiateJoinGroup`:

```
             public void onFailure(RuntimeException e) {
                    // we handle failures below after the request finishes. if the join completes
                    // after having been woken up, the exception is ignored and we will rejoin
                    synchronized (AbstractCoordinator.this) {
                        state = MemberState.UNJOINED;
                    }
                }
```

I.e. it will also be swallowed and retry. Should we instead let this consumer to fail by throwing this exception all the way up, like in KafkaProducer we throw `ProduceFencedException` all the way to the caller's face.

Reply:
I could see #initiateJoinGroup being called and we do check future later here for exceptions:
```
} else {
                resetJoinGroupFuture();
                final RuntimeException exception = future.exception();
                if (exception instanceof UnknownMemberIdException ||
                        exception instanceof RebalanceInProgressException ||
                        exception instanceof IllegalGenerationException ||
                        exception instanceof MemberIdRequiredException)
                    continue;
                else if (!future.isRetriable())
                    throw exception;
```
So my understanding is that we will not retry in the case, since the exception is thrown.

Reply:
Ah makes sense.